### DISCUSSION TITLE
server_example: Read MyVariable using namespace and string identifier

### DISCUSSION THREAD

**Discussion by mahadi** [FreeOpcUa/opcua-asyncio#1235] [Q&A] [answered]
I am trying to read the "MyVariable" node from server_example .py using (sync) client_example.py.
The examples themselves are working fine, but i ask myself if it shouldnt be possible to read the node also using its name

``` python
client.get_node("ns=2;s=MyVariable").read_browse_name()

> BadNodeIdUnknown: "The node id refers to a node that does not exist in the server address space."(BadNodeIdUnknown)
```

If i recursively traverse the tree i get (the same as UaExpert):

``` python
{QualifiedName(NamespaceIndex=2, Name='MyObject'): [
    {QualifiedName(NamespaceIndex=2, Name='MyVariable'): SyncNode(NodeId(Identifier=13, NamespaceIndex=2, NodeIdType=<NodeIdType.FourByte: 1>))},
    ...
]}
```
And i know from WAGO PLCs that it is possible to read nodes using this method.

Did i miss something here? Or is there a special reason why it doesnt work in the examples?

**Comment by AndreasHeine** [comment]
looks like ns=2;s=MyVariable does not exist in your ua server!

BadNodeIdUnknown is a response from the server!? maybe the ns-idx wrong?

**Reply by mahadi** [reply]
It should because in UaExpert, it is visible:
![grafik](https://user-images.githubusercontent.com/124618/222742583-54706327-a828-49eb-91fe-01e040ff972d.png)
![grafik](https://user-images.githubusercontent.com/124618/222742742-6ddc6c68-0d4c-425d-aa13-6cec4bec6b08.png)

**Reply by schroeder-** [reply]
NodeId is ns=2;i=13

**Comment by schroeder-** [comment, marked as answer]
nodeid != browsename try this:
client.get_node("ns=2;i=13")

**Reply by mahadi** [reply]
Is it possible to populate the nodeid with the browsename in the server?

``` python
myobj = await server.nodes.objects.add_object(idx, "MyObject")
myvar = await myobj.add_variable(idx, "MyVariable", 6.7)
```
This is how it is initialized in the server example.

**Reply by AndreasHeine** [reply]
do you mean the IdentifierType Sting? with s=MaVariable? thats possible!

**Reply by mahadi** [reply]
Yes, this is what i mean.

**Reply by mahadi** [reply]
Is there a short example somewhere how to do this?

**Reply by schroeder-** [reply]
```Python
myobj = await server.nodes.objects.add_object(NodeId(f'ns={idx};s=MyObject'), "MyObject")
myvar = await myobj.add_variable(NodeId(f'ns={idx};s=MyVariable'), "MyVariable", 6.7)
```
or 
```Python
myobj = await server.nodes.objects.add_object(NodeId('MyObject', idx), "MyObject")
myvar = await myobj.add_variable(NodeId('MyVariable', idx), "MyVariable", 6.7)
```

**Reply by mahadi** [reply]
Ah great, thanks a lot to both of you for the quick response!

**Reply by mahadi** [reply]
Btw: For me only the first option worked:

``` python
    mynsobj = await server.nodes.objects.add_object(ua.NodeId(f'ns={idx};s=MyObjectNs'), "MyObjectNs")
    mynsvar = await mynsobj.add_variable(ua.NodeId(f'ns={idx};s=MyVariableNs'), "MyVariableNs", 6.1)
```
For the second one 

``` python
    mynsobj = await server.nodes.objects.add_object(ua.NodeId(idx, "MyObjectNs"), "MyObjectNs")
    mynsvar = await mynsobj.add_variable(ua.NodeId(idx, "MyVariableNs"), "MyVariableNs", 6.1)
```

i got:

``` python
  File "opcua-asyncio/examples/server-example.py", line 87, in main
    mynsobj = await server.nodes.objects.add_object(ua.NodeId(idx, "MyObjectNs"), "MyObjectNs")
  File "<string>", line 6, in __init__
  File "lib/python3.10/site-packages/asyncua/ua/uatypes.py", line 414, in __post_init__
    elif self.Identifier < 2 ** 16 and self.NamespaceIndex < 255:
TypeError: '<' not supported between instances of 'str' and 'int'
```

**Reply by schroeder-** [reply]
My bad wrong argument order:
```Python
myobj = await server.nodes.objects.add_object(NodeId('MyObject', idx), "MyObject")
myvar = await myobj.add_variable(NodeId('MyVariable', idx), "MyVariable", 6.7)
```
